homeassistant/components/bluetooth/manager.py

```diff
@@ -330,28 +317,32 @@ def async_register_callback(
         matcher: BluetoothCallbackMatcher | None,
     ) -> Callable[[], None]:
         """Register a callback."""
+        callback_matcher = BluetoothCallbackMatcherWithCallback(callback=callback)
         if not matcher:
-            matcher = BluetoothCallbackMatcher(connectable=True)
-        if CONNECTABLE not in matcher:
-            matcher[CONNECTABLE] = True
-        connectable = matcher[CONNECTABLE]
+            callback_matcher[CONNECTABLE] = True
+        else:
+            # We could write out every item in the typed dict here
+            # but that would be a bit inefficient and verbose.
+            callback_matcher.update(matcher)  # type: ignore[typeddict-item]
+            callback_matcher[CONNECTABLE] = matcher.get(CONNECTABLE, True)
 
-        callback_entry = (callback, matcher)
-        callbacks = self._get_callbacks_by_type(connectable)
-        callbacks.append(callback_entry)
+        connectable = callback_matcher[CONNECTABLE]
+        self._callback_index.add_with_address(callback_matcher)
+        self._callback_index.build()
 
         @hass_callback
         def _async_remove_callback() -> None:
-            callbacks.remove(callback_entry)
+            self._callback_index.remove_with_address(callback_matcher)
+            self._callback_index.build()
```

Comment:
Isn't it weird that we should call build manually after removal? Could we mark it as "build necessary" and do it when the next call comes in ?

Reply:
Its also used for the integration matchers and we only want to build it once after we add them all.

Reply:
But we subclass it so we could build in the *_with_address

Reply:
ab4ff858941299c619c591665dd19a70dbd6e569